### PR TITLE
WIP: Auto-Import User-Defined Cosmology Realizations

### PR DESCRIPTION
**Pull request by nstarman**
### Description

Allow the user to add an ECSV file of custom realizations to a config directory (default in `~/.astropy/cosmologies/`) and have them be created at import.  Also, I think using the ECSV format will eventually enable cool things like working through an MCMC chain of cosmologies.

For backward compatibility, the file is read and dicts of each row added into ``parameters.py``.

Part of #10673 
Blocked by ~#11530~ #11570 ~#11542~  #11559

@embray, this PR relates to our convo on non-entrypoint custom realizations.

- - -
OLD Description

> Instead of storing parameters as dictionaries in the file ``parameters``, the parameters for built-in realizations are kept in an ECSV. This enables easy comparison of realizations and, more importantly, lays the foundation for the next PR, which is allow the user to add an ECSV file of custom realizations to the astropy config directory and have them be created at import. 
Also, I think using the ECSV format will eventually enable cool things like working through an MCMC chain of cosmologies.
> For backward compatibility, the file is read and dicts of each row added into ``parameters.py``.

### CODE REVIEW

**Comment by nstarman** [comment]
Uh oh! I forgot PyYAML is not a run-time dependency... how can an ECSV file be read?

**Comment by taldcroft** [comment]
@nstarman - unfortunately I can't think of a way to use ECSV for this case. It really does require YAML, so the choice is either don't use ECSV or make YAML a requirement for `cosmology` (and hence `astropy`).

BTW, if you are wondering about why we used YAML not JSON for ECSV.. I often think about this myself:
- A key showstopper is that JSON does not allow numeric values for a dict key. So it cannot represent arbitrary pure-Python data structures.
- Much better system for extensibility.
- More human-readable.

I spent a little time thinking about ways of reading key bits of the YAML header "by hand" without YAML, but at the end I don't think this will be worth it. Sigh.

**Comment by taldcroft** [comment]
BTW, 👍 on the idea, even if it ends up not working out!

**Comment by nstarman** [comment]
> BTW, 👍 on the idea, even if it ends up not working out!

Thanks!

>  unfortunately I can't think of a way to use ECSV for this case. It really does require YAML, so the choice is either don't use ECSV or make YAML a requirement for cosmology (and hence astropy).

Well, I don't want to add any requirements... So I guess I'll use JSON for astropy's internal storage 😢 .
I think I'll add class methods ``Cosmology.from/to_json`` that for JSON I/O.

For user-defined cosmologies, I think the ECSV is more logical and the code I have here will work well. Perhaps I'll just transition this PR into user-defined cosmologies. 

Thanks @taldcroft for the quick reply.

**Comment by pllim** [comment]
> I guess I'll use JSON

What about FITS? 😆

**Comment by nstarman** [comment]
I'm also planning on [Linear A](https://en.wikipedia.org/wiki/Linear_A) as soon as it's deciphered.

File: setup.cfg [hunk 1]
```diff
@@ -93,6 +93,7 @@ docs =
 [options.package_data]
 * = data/*, data/*/*, data/*/*/*, data/*/*/*/*, data/*/*/*/*/*, data/*/*/*/*/*/*
 astropy = astropy.cfg, CITATION
+astropy.cosmology = realizations.ecsv
```
Comment (mhvk) [inline setup.cfg:96]: The file itself does not seem to be included in the PR. Also, a possible problem if that becomes our primary source: we would then need to depend on YAML.

Comment (nstarman) [reply setup.cfg]: Thanks! that's actually a relic of the old PR direction. I've removed it.  
I now have a working implementation of auto-import of all user-defined realizations in any file matching the glob `~/.astropy/cosmology/*.ecsv`.

Screen Shot 2021-04-13 at 21 33 11

Comment (nstarman) [reply setup.cfg]: I've added WIP to the title, until the spaghetti code is cleaned up.

**Comment by embray** [comment]
@nstarman At some point I'd like to have a discussion to better wrap my head around your cosmology plans.  I like where this is headed though and I think you have the right idea.  +1 to using YAML to describe cosomologies, and if we need to add pyyaml as a dependency for astropy I would not be opposed.

**Comment by nstarman** [comment]
@embray Adding PyYAML would be awesome. @taldcroft and I had some discussion in a different PR about this... and it would be very nice to have a clear way of serializing quantities.
I would be more than happy to chat about my cosmology plans... perhaps that would help me understand them as well! 😆. I've defined some broad goals, but as you can see from the various WIP PRs, I haven't yet coordinated a specific line of attack. Would this be best done on Slack? There's the languishing cosmology channel.

**Comment by nstarman** [comment]
Worth doing, but this was way outdated and the wrong approach.
